designer/server/src/main/scala/pl/touk/nussknacker/ui/db/timeseries/questdb/QuestDbFEStatisticsRepository.scala

```diff
@@ -0,0 +1,274 @@
+// 9. multi tenant metrics -> add todo that if there are many instances we should change diagrams on grafana
+// 10. flush data in close?
+// todo polish this class in next commits
+private class QuestDbFEStatisticsRepository(private val engine: AtomicReference[CairoEngine], private val clock: Clock)(
+    private implicit val ec: ExecutionContextExecutorService
+) extends FEStatisticsRepository[Future]
+    with LazyLogging {
+
+  private val sqlContext = new ThreadAwareObjectPool(() =>
+    new SqlExecutionContextImpl(engine.get(), 1).`with`(AllowAllSecurityContext.INSTANCE, null)
+  )
+
+  private val recordCursorPool = new ThreadAwareObjectPool(() => engine.get().select(selectQuery, sqlContext.get()))
+
+  private val walTableWriterPool = new ThreadAwareObjectPool(() =>
+    engine.get().getWalWriter(engine.get().getTableTokenIfExists(tableName))
+  )
+
+  private val flushDataTask      = new AtomicReference(createFlushDataTask())
+  private lazy val retentionTask = new AtomicReference(createRetentionTask())
+  private val shouldCleanUpData  = new AtomicBoolean(false)
+
+  override def write(statistics: Map[String, Long]): Future[Unit] = withExceptionHandling(() => {
+    val statsWalTableWriter = walTableWriterPool.get()
+    statistics.foreach { entry =>
+      val row = statsWalTableWriter.newRow(currentTimeMicros())
+      row.putStr(0, entry._1)
+      row.putLong(1, entry._2)
+      row.append()
+    }
+    statsWalTableWriter.commit()
+  })
+
+  override def read(): Future[Map[String, Long]] = withExceptionHandling(() => {
+    recordCursorPool
+      .get()
+      .fetch(sqlContext.get()) { record =>
+        val name  = record.getStrA(0).toString
+        val count = record.getLong(1)
+        name -> count
+      }
+      .toMap
+  })
+
+  private def currentTimeMicros(): Long =
+    Math.multiplyExact(clock.instant().toEpochMilli, 1000L)
+
+  private def close(): Unit = {
+    recordCursorPool.clear()
+    walTableWriterPool.clear()
+    sqlContext.clear()
+    flushDataTask.get().close()
+    retentionTask.get().close()
+  }
+
+  private def initialize(): Unit = {
+    engine.get().ddl(createTableQuery, sqlContext.get())
+  }
+
+  private def withExceptionHandling[T](dbAction: () => T): Future[T] =
+    Future {
+      if (engine.get().tableExists(tableName)) {
+        runDbAction(dbAction)
+      } else {
+        recoverWithRetry(new IllegalStateException("Table does not exist"), dbAction).get
+      }
+    }
+
+  private def runDbAction[T](dbAction: () => T) = {
+    Try(dbAction()).recoverWith {
+      case ex: CairoException if ex.isCritical || ex.isTableDropped =>
+        recoverWithRetry(ex, dbAction)
+      case ex: SqlException if ex.getMessage.contains("table does not exist") =>
+        recoverWithRetry(ex, dbAction)
+      case ex =>
+        logger.warn("DB exception", ex)
+        Failure(ex)
+    }.get
+  }
+
+  private def recoverWithRetry[T](ex: Exception, dbAction: () => T): Try[T] = {
```

Comment:
In next commit I will try to remove the table existence check. For now I left as it is and I added todo to resolve the problem later.
2. and 3. it has been done because write action in some circumstances works as buffer and write action will not throw exception even if table does not exist. Table existence is checked in catch section because in read operations table existence is cached in memory and verification is done on actual db call.

Reply:
https://github.com/TouK/nussknacker/pull/6220